pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -375,37 +406,51 @@ func (r *ReconcileClusterDisruption) updateNoout(clusterInfo *cephclient.Cluster
 	return nil
 }
 
-func (r *ReconcileClusterDisruption) getOSDFailureDomains(clusterInfo *cephclient.ClusterInfo, request reconcile.Request, poolFailureDomain string) ([]string, []string, error) {
+func (r *ReconcileClusterDisruption) getOSDFailureDomains(clusterInfo *cephclient.ClusterInfo, request reconcile.Request, poolFailureDomain string) ([]string, []string, []string, error) {
 	osdDeploymentList := &appsv1.DeploymentList{}
 	namespaceListOpts := client.InNamespace(request.Namespace)
 	topologyLocationLabel := fmt.Sprintf(osd.TopologyLocationLabel, poolFailureDomain)
 	err := r.client.List(context.TODO(), osdDeploymentList, client.MatchingLabels{k8sutil.AppAttr: osd.AppName}, namespaceListOpts)
 	if err != nil {
-		return []string{}, []string{}, errors.Wrap(err, "failed to list osd deployments")
+		return nil, nil, nil, errors.Wrap(err, "failed to list osd deployments")
 	}
 
 	allFailureDomains := sets.NewString()
-	drainingFailureDomains := sets.NewString()
+	nodeDrainFailureDomains := sets.NewString()
+	osdDownFailureDomains := sets.NewString()
 
 	for _, deployment := range osdDeploymentList.Items {
 		labels := deployment.Spec.Template.ObjectMeta.GetLabels()
 		failureDomainName := labels[topologyLocationLabel]
 		if failureDomainName == "" {
-			return []string{}, []string{}, errors.Errorf("failed to get the topology location label %q in OSD deployment %q",
+			return nil, nil, nil, errors.Errorf("failed to get the topology location label %q in OSD deployment %q",
 				topologyLocationLabel, deployment.Name)
 		}
 
+		// Assume node drain if osd deployment ReadyReplicas count is 0 and OSD pod is not scheduled on a node
 		if deployment.Status.ReadyReplicas < 1 {
-			if !drainingFailureDomains.Has(failureDomainName) {
-				drainingFailureDomains.Insert(failureDomainName)
+			if !osdDownFailureDomains.Has(failureDomainName) {
+				osdDownFailureDomains.Insert(failureDomainName)
 			}
-		} else {
-			if !allFailureDomains.Has(failureDomainName) {
-				allFailureDomains.Insert(failureDomainName)
+			isDrained, err := hasOSDNodeDrained(r.client, request.Namespace, labels[osd.OsdIdLabelKey])
+			if err != nil {
+				return nil, nil, nil, errors.Wrapf(err, "failed to check if osd %q node is drained", deployment.Name)
+			}
+			if isDrained {
+				logger.Infof("osd %q is down and a possible node drain is detected", deployment.Name)
+				if !nodeDrainFailureDomains.Has(failureDomainName) {
```

Comment:
Do you need to call `Has()`? If it's a set, you should be able to insert the same element and not have a duplicate. Rook has a `util.Set` type that would behave this way.

Reply:
`util.Set` does not provide a way to get sorted slice of strings. So using `List()` method in `"k8s.io/apimachinery/pkg/util/sets"`

Reply:
One option would be to add `ToSortedSlice()` in `util.Set` and use that. I would prefer to do that in a follow up PR.

Reply:
Oh if you need to sort the items, don't use `util.Set`. It's just for an unordered Set since it's based on a map. Remind me, why do we need to sort the failure domains?

Reply:
we get the list of draining failure domain on every reconcile and then select from failure domain in that list as the actual draining failure domain ([here](https://github.com/rook/rook/blob/4996b1cbb9fec7795651efda01d409a333762374/pkg/operator/ceph/disruption/clusterdisruption/osd.go#L226)) . So we might want this list to be sorted to get the same failure domain on every reconcile.